File: protocol/protocol.py
import struct
import json
from enum import IntEnum
from dataclasses import dataclass
from typing import Optional

MAGIC_NUMBER = 0x4157534D

class PacketType(IntEnum):
    HANDSHAKE_REQUEST = 0x01
    HANDSHAKE_RESPONSE = 0x02
    VIDEO_FRAME = 0x10
    AUDIO_FRAME = 0x11
    CONTROL_COMMAND = 0x20
    CONTROL_RESPONSE = 0x21
    KEEPALIVE = 0x30
    ERROR = 0xFF

@dataclass
class PacketHeader:
    magic: int
    packet_type: PacketType
    payload_size: int
    timestamp: int
    sequence_number: int
    
    HEADER_SIZE = 21
    
    def to_bytes(self) -> bytes:
        return struct.pack(
            '>IBIQI',
            self.magic,
            self.packet_type,
            self.payload_size,
            self.timestamp,
            self.sequence_number
        )
    
    @staticmethod
    def from_bytes(data: bytes) -> 'PacketHeader':
        magic, packet_type, payload_size, timestamp, seq_num = struct.unpack('>IBIQI', data)
        return PacketHeader(
            magic=magic,
            packet_type=PacketType(packet_type),
            payload_size=payload_size,
            timestamp=timestamp,
            sequence_number=seq_num
        )

class Packet:
    def __init__(self, packet_type: PacketType, payload: bytes, timestamp: int = 0, seq_num: int = 0):
        self.header = PacketHeader(
            magic=MAGIC_NUMBER,
            packet_type=packet_type,
            payload_size=len(payload),
            timestamp=timestamp,
            sequence_number=seq_num
        )
        self.payload = payload
    
    def to_bytes(self) -> bytes:
        return self.header.to_bytes() + self.payload
    
    @staticmethod
    def from_bytes(data: bytes) -> 'Packet':
        header = PacketHeader.from_bytes(data[:PacketHeader.HEADER_SIZE])
        payload = data[PacketHeader.HEADER_SIZE:PacketHeader.HEADER_SIZE + header.payload_size]
        packet = Packet(header.packet_type, payload)
        packet.header = header
        return packet

class ControlCommand:
    @staticmethod
    def create_command(command: str, parameters: dict) -> str:
        return json.dumps({
            "command": command,
            "parameters": parameters
        })
    
    @staticmethod
    def parse_command(json_str: str) -> tuple:
        data = json.loads(json_str)
        return data["command"], data["parameters"]
    
    @staticmethod
    def set_resolution(width: int, height: int) -> str:
        return ControlCommand.create_command("set_resolution", {"width": width, "height": height})
    
    @staticmethod
    def set_fps(fps: int) -> str:
        return ControlCommand.create_command("set_fps", {"fps": fps})
    
    @staticmethod
    def switch_camera(camera: str) -> str:
        return ControlCommand.create_command("switch_camera", {"camera": camera})
    
    @staticmethod
    def set_flash(enabled: bool) -> str:
        return ControlCommand.create_command("set_flash", {"enabled": enabled})
    
    @staticmethod
    def set_focus(mode: str, distance: float = 0.0) -> str:
        return ControlCommand.create_command("set_focus", {"mode": mode, "distance": distance})
    
    @staticmethod
    def set_brightness(value: float) -> str:
        return ControlCommand.create_command("set_brightness", {"value": value})

File: protocol/Protocol.kt
package com.streamapp.protocol

import org.json.JSONObject
import java.nio.ByteBuffer
import java.nio.ByteOrder

const val MAGIC_NUMBER = 0x4157534D

enum class PacketType(val value: Byte) {
    HANDSHAKE_REQUEST(0x01),
    HANDSHAKE_RESPONSE(0x02),
    VIDEO_CONFIG(0x03),
    AUDIO_CONFIG(0x04),
    VIDEO_FRAME(0x10),
    AUDIO_FRAME(0x11),
    CONTROL_COMMAND(0x20),
    CONTROL_RESPONSE(0x21),
    KEEPALIVE(0x30),
    ERROR(0xFF.toByte());
    
    companion object {
        fun fromByte(byte: Byte): PacketType {
            return values().find { it.value == byte } ?: ERROR
        }
    }
}

data class PacketHeader(
    val magic: Int,
    val packetType: PacketType,
    val payloadSize: Int,
    val timestamp: Long,
    val sequenceNumber: Int
) {
    companion object {
        const val HEADER_SIZE = 21
        
        fun fromBytes(data: ByteArray): PacketHeader {
            val buffer = ByteBuffer.wrap(data).order(ByteOrder.BIG_ENDIAN)
            return PacketHeader(
                magic = buffer.int,
                packetType = PacketType.fromByte(buffer.get()),
                payloadSize = buffer.int,
                timestamp = buffer.long,
                sequenceNumber = buffer.int
            )
        }
    }
    
    fun toBytes(): ByteArray {
        val buffer = ByteBuffer.allocate(HEADER_SIZE).order(ByteOrder.BIG_ENDIAN)
        buffer.putInt(magic)
        buffer.put(packetType.value)
        buffer.putInt(payloadSize)
        buffer.putLong(timestamp)
        buffer.putInt(sequenceNumber)
        return buffer.array()
    }
}

class Packet(
    val packetType: PacketType,
    val payload: ByteArray,
    timestamp: Long = System.currentTimeMillis(),
    sequenceNumber: Int = 0
) {
    val header = PacketHeader(
        magic = MAGIC_NUMBER,
        packetType = packetType,
        payloadSize = payload.size,
        timestamp = timestamp,
        sequenceNumber = sequenceNumber
    )
    
    fun toBytes(): ByteArray {
        return header.toBytes() + payload
    }
    
    companion object {
        fun fromBytes(data: ByteArray): Packet {
            val header = PacketHeader.fromBytes(data.sliceArray(0 until PacketHeader.HEADER_SIZE))
            val payload = data.sliceArray(PacketHeader.HEADER_SIZE until PacketHeader.HEADER_SIZE + header.payloadSize)
            return Packet(header.packetType, payload, header.timestamp, header.sequenceNumber)
        }
    }
}

object ControlCommand {
    fun createCommand(command: String, parameters: Map<String, Any>): String {
        val json = JSONObject()
        json.put("command", command)
        json.put("parameters", JSONObject(parameters))
        return json.toString()
    }
    
    fun parseCommand(jsonStr: String): Pair<String, Map<String, Any>> {
        val json = JSONObject(jsonStr)
        val command = json.getString("command")
        val params = json.getJSONObject("parameters")
        val paramsMap = mutableMapOf<String, Any>()
        params.keys().forEach { key ->
            paramsMap[key] = params.get(key)
        }
        return Pair(command, paramsMap)
    }
    
    fun setResolution(width: Int, height: Int): String {
        return createCommand("set_resolution", mapOf("width" to width, "height" to height))
    }
    
    fun setFps(fps: Int): String {
        return createCommand("set_fps", mapOf("fps" to fps))
    }
    
    fun switchCamera(camera: String): String {
        return createCommand("switch_camera", mapOf("camera" to camera))
    }
    
    fun setFlash(enabled: Boolean): String {
        return createCommand("set_flash", mapOf("enabled" to enabled))
    }
    
    fun setFocus(mode: String, distance: Float = 0f): String {
        return createCommand("set_focus", mapOf("mode" to mode, "distance" to distance))
    }
    
    fun setBrightness(value: Float): String {
        return createCommand("set_brightness", mapOf("value" to value))
    }
    
    fun setZoom(ratio: Float): String {
        return createCommand("set_zoom", mapOf("ratio" to ratio))
    }
    
    fun setManualMode(enabled: Boolean): String {
        return createCommand("set_manual_mode", mapOf("enabled" to enabled))
    }
    
    fun setManualISO(iso: Int): String {
        return createCommand("set_manual_iso", mapOf("iso" to iso))
    }
    
    fun setManualShutter(nanos: Long): String {
        return createCommand("set_manual_shutter", mapOf("nanos" to nanos))
    }
    
    fun setManualFocus(distance: Float): String {
        return createCommand("set_manual_focus", mapOf("distance" to distance))
    }
    
    fun setWhiteBalance(mode: String, kelvin: Int): String {
        return createCommand("set_white_balance", mapOf("mode" to mode, "kelvin" to kelvin))
    }
    
    fun setTapFocus(x: Float, y: Float, screenWidth: Int, screenHeight: Int): String {
        return createCommand("set_tap_focus", mapOf(
            "x" to x,
            "y" to y,
            "screenWidth" to screenWidth,
            "screenHeight" to screenHeight
        ))
    }
    
    fun getCameraCapabilities(): String {
        return createCommand("get_camera_capabilities", emptyMap())
    }
}

